File: reference/scenarios.mjs
import assert from 'node:assert/strict';

const gate = () => Promise.withResolvers();
const fiber = value => value.ctx.fiber; // unwrap the PromiseLike proxy; retain the original Fiber
const P = (apply, inject = []) => ({ apply, inject });
const guard = (promise, id) => {
  let timer;
  return Promise.race([promise, new Promise((_, reject) => {
    timer = setTimeout(() => reject(new Error(`Scenario timed out: ${id}`)), 15000);
  })]).finally(() => clearTimeout(timer));
};
class Messages {
  handlers = new Set();
  subscribe(handler) { this.handlers.add(handler); return () => this.handlers.delete(handler); }
  emit(name) { for (const h of [...this.handlers]) h(name); }
  get count() { return this.handlers.size; }
}

export async function runScenarios(Context) {
  const cases = [];
  async function run(id, body) {
    const ctx = new Context();
    const trace = [];
    try {
      await guard(body(ctx, trace), id);
      cases.push({ id, trace: [...trace] });
      console.error(`PASS ${id}`);
    } finally { await guard(ctx.fiber.dispose(), `${id}/cleanup`); }
  }
  await run('C01-greeting-lifecycle', async (ctx, log) => {
    const a = new Messages(), b = new Messages();
    const plugin = P((c, prefix) => {
      const source = c.get('messages');
      c.effect(() => source.subscribe(name => log.push(`${prefix}, ${name}!`)));
    }, ['messages']);
    const f = fiber(ctx.plugin(plugin, 'Hello'));
    await f.await(); assert.equal(f.state, 0); log.push('pending');
    const first = ctx.provide('messages', a);
    await f.await(); assert.equal(a.count, 1); a.emit('Otto');
    f.update('Hi'); await f.await(); assert.equal(a.count, 1); a.emit('Otto');
    await first(); assert.equal(a.count, 0); assert.equal(f.state, 0); log.push('provider-removed');
    const second = ctx.provide('messages', b);
    await f.await(); b.emit('Otto'); await f.dispose();
    assert.equal(b.count, 0); b.emit('ignored'); await second(); log.push('disposed');
  });
  await run('C02-effect-immediate-single-shot', async (ctx, log) => {
    const e = ctx.effect(() => { log.push('setup'); return () => log.push('cleanup'); });
    log.push('returned'); await e; await e(); await e(); log.push('disposed-twice');
  });
  await run('C03-reentrant-owner-disposal', async (ctx, log) => {
    let stopping;
    ctx.effect(() => {
      log.push('setup-start'); stopping = ctx.fiber.dispose(); log.push('setup-end');
      return () => log.push('cleanup');
    });
    log.push('returned'); await stopping; assert.equal(ctx.fiber.state, 2); log.push('root-active');
  });
  await run('C04-pending-owned-effect', async (ctx, log) => {
    const f = fiber(ctx.plugin(P(() => { throw new Error('must not apply'); }, ['absent'])));
    f.ctx.effect(() => { log.push('setup'); return () => log.push('cleanup'); });
    await f.await(); assert.equal(f.state, 0); await f.dispose(); assert.equal(f.state, 4); log.push('disposed');
  });
  await run('C05-dispose-before-load-checkpoint', async (ctx, log) => {
    const f = fiber(ctx.plugin(P(() => { log.push('must-not-apply'); })));
    await f.dispose(); assert.equal(f.state, 4); assert.equal(log.length, 0); log.push('disposed-before-apply');
  });
  await run('C06-async-setup-and-cleanup', async (ctx, log) => {
    const setup = gate(), cleanup = gate(), entered = gate();
    const e = ctx.effect(async () => {
      log.push('setup-start'); await setup.promise; log.push('setup-end');
      return async () => { log.push('cleanup-start'); entered.resolve(); await cleanup.promise; log.push('cleanup-end'); };
    });
    try {
      let stopped = false;
      const stopping = ctx.fiber.dispose().then(() => { stopped = true; });
      assert.equal(stopped, false); log.push('waits-for-setup'); setup.resolve();
      await entered.promise; assert.equal(stopped, false); log.push('waits-for-cleanup');
      cleanup.resolve(); await stopping; await e; log.push('stopped');
    } finally { setup.resolve(); cleanup.resolve(); }
  });
  await run('C07-owner-joins-started-cleanup', async (ctx, log) => {
    const entered = gate(), release = gate();
    const e = ctx.effect(() => async () => {
      log.push('cleanup-start'); entered.resolve(); await release.promise; log.push('cleanup-end');
    });
    try {
      const first = e(); await entered.promise; await e(); log.push('second-public-call-returned');
      let stopped = false;
      const stop = ctx.fiber.dispose().then(() => { stopped = true; });
      assert.equal(stopped, false); log.push('owner-waits'); release.resolve();
      await first; await stop; log.push('stopped');
    } finally { release.resolve(); }
  });
  await run('C08-nested-effect-reverse-order', async (ctx, log) => {
    const outer = ctx.effect(() => [
      ctx.effect(() => { log.push('setup-a'); return () => log.push('cleanup-a'); }),
      ctx.effect(() => { log.push('setup-b'); return () => log.push('cleanup-b'); }),
    ]);
    await outer(); await ctx.fiber.dispose();
    assert.deepEqual(log, ['setup-a', 'setup-b', 'cleanup-b', 'cleanup-a']);
  });
  await run('C09-generator-setup-rollback', async (ctx, log) => {
    assert.throws(() => ctx.effect(function* () {
      yield ctx.effect(() => { log.push('setup-a'); return () => log.push('cleanup-a'); });
      yield ctx.effect(() => { log.push('setup-b'); return () => log.push('cleanup-b'); });
      throw new Error('setup-failed');
    }), /setup-failed/);
    await ctx.fiber.dispose(); log.push('failure-observed');
  });
  await run('C10-failed-group-does-not-starve-sibling', async (ctx, log) => {
    ctx.effect(() => () => log.push('sibling'));
    ctx.effect(() => [() => log.push('must-not-run'), () => { log.push('failed-cleanup'); throw new Error('cleanup'); }]);
    await ctx.fiber.dispose();
    assert.deepEqual(log, ['failed-cleanup', 'sibling']);
    assert.equal(ctx.logger.buffer.filter(m => m.type === 'error').length, 1);
    log.push('one-diagnostic');
  });
  await run('C11-reject-effect-during-unload', async (ctx, log) => {
    ctx.effect(() => () => {
      assert.throws(() => ctx.effect(() => { log.push('must-not-setup'); return () => {}; }),
        error => error.code === 'INACTIVE_EFFECT');
      log.push('rejected');
    });
    await ctx.fiber.dispose(); assert.equal(log.length, 1);
  });
  await run('C12-apply-failure-cleanup', async (ctx, log) => {
    const f = fiber(ctx.plugin(P(c => {
      c.effect(() => { log.push('setup'); return () => log.push('cleanup'); }); throw new Error('apply-failed');
    })));
    await assert.rejects(f.await(), /apply-failed/); assert.equal(f.state, 3); log.push('failed');
    await f.dispose(); assert.equal(f.state, 4); log.push('disposed');
  });
  await run('C13-loading-provider-invisible', async (ctx, log) => {
    const published = gate(), release = gate(), value = {};
    const consumer = fiber(ctx.plugin(P(() => { log.push('consumer-active'); }, ['service'])));
    const provider = fiber(ctx.plugin(P(async c => { c.provide('service', value); published.resolve(); await release.promise; })));
    try {
      await published.promise; assert.equal(provider.state, 1); assert.equal(ctx.get('service'), undefined);
      assert.equal(ctx.get('service', false), value); await consumer.await(); assert.equal(consumer.state, 0);
      log.push('loading-not-visible'); release.resolve(); await provider.await(); await consumer.await();
      assert.equal(consumer.state, 2); log.push('ready');
    } finally { release.resolve(); }
  });
  await run('C14-inject-child-not-parent-restart', async (ctx, log) => {
    let child, parents = 0;
    const parent = fiber(ctx.plugin(P(c => {
      parents++;
      child = fiber(c.inject(['service'], cc => {
        cc.effect(() => { log.push('child-setup'); return () => log.push('child-cleanup'); });
      }));
    })));
    await parent.await(); assert.equal(child.state, 0);
    const service = ctx.provide('service', {}); await child.await(); await service();
    assert.equal(parents, 1); assert.equal(parent.state, 2); assert.equal(child.state, 0);
    log.push('parent-unchanged'); await parent.dispose(); assert.equal(child.state, 4);
    log.push('child-disposed-with-parent');
  });
  await run('C15-pending-and-active-config-update', async (ctx, log) => {
    let validations = 0;
    const plugin = {
      inject: ['service'],
      Config: { '~standard': { version: 1, vendor: 'cordis-slice-probe', validate(value) {
        validations++; return value === 'good' ? { value } : { issues: [{ message: 'expected good' }] };
      }}},
      apply: (_c, config) => { log.push('apply:' + config); },
    };
    const f = fiber(ctx.plugin(plugin, 'bad')); await f.await(); assert.equal(validations, 0);
    f.update('good'); assert.equal(validations, 0); log.push('deferred-validation');
    ctx.provide('service', {}); await f.await();
    assert.throws(() => f.update('bad')); assert.equal(f.config, 'good'); assert.equal(f._config, 'bad');
    assert.equal(f.state, 2); log.push('invalid-update-preserves-active-config');
    await assert.rejects(f.restart()); assert.equal(f.state, 3); log.push('restart-uses-raw-config');
    f.update('good'); await f.await(); assert.equal(f.state, 2); log.push('recovered');
  });
  await run('C16-multiple-dependencies', async (ctx, log) => {
    let starts = 0;
    const f = fiber(ctx.plugin(P(() => { starts++; }, ['a', 'b'])));
    const a = ctx.provide('a', {}); await f.await(); assert.equal(f.state, 0);
    ctx.provide('b', {}); await f.await(); assert.equal(starts, 1);
    await a(); assert.equal(f.state, 0); ctx.provide('a', {}); await f.await(); assert.equal(starts, 2);
    log.push('both-required;two-activations');
  });
  await run('C17-owner-joins-disposing-child', async (ctx, log) => {
    const entered = gate(), release = gate();
    const child = fiber(ctx.plugin(P(c => { c.effect(() => async () => {
      log.push('child-cleanup-start'); entered.resolve(); await release.promise; log.push('child-cleanup-end');
    }); })));
    try {
      await child.await(); const first = child.dispose(); await entered.promise;
      let stopped = false; const rootStop = ctx.fiber.dispose().then(() => { stopped = true; });
      assert.equal(stopped, false); log.push('root-waits'); release.resolve();
      await first; await rootStop; log.push('stopped');
    } finally { release.resolve(); }
  });
  await run('C18-stable-context-on-restart', async (ctx, log) => {
    const seen = [];
    const f = fiber(ctx.plugin(P(c => { seen.push(c); })));
    await f.await(); const uid = f.uid; await f.restart();
    assert.equal(seen.length, 2); assert.equal(seen[0], seen[1]); assert.equal(f.uid, uid);
    log.push('same-context;same-fiber;two-applies');
  });
  // Five named tests adapted from the upstream dispose.spec.ts; assertions retained.
  await run('U01-effects-dispose-by-plugin', async (root, log) => {
    let calls = 0;
    const f = fiber(root.plugin(P(ctx => { ctx.effect(() => () => { calls++; }, 'test'); })));
    await f.await();
    assert.deepEqual(f.getEffects(), [{ label: 'test', children: [] }]);
    assert.equal(calls, 0); await f.dispose(); assert.equal(calls, 1);
    await f.dispose(); assert.equal(calls, 1); log.push('metadata:test;calls:0,1,1');
  });
  await run('U02-effects-dispose-manually', async (root, log) => {
    let calls = 0;
    const e = root.effect(() => () => { calls++; });
    assert.deepEqual(root.fiber.getEffects(), [{ label: 'anonymous', children: [] }]);
    assert.equal(calls, 0); const first = e(); assert.equal(calls, 1);
    const second = e(); assert.equal(calls, 1); await first; await second;
    log.push('metadata:anonymous;calls:0,1,1');
  });
  await run('U03-effects-return-with-error', async (root, log) => {
    const sequence = [];
    assert.throws(() => root.effect(() => { throw new Error('test'); }), /test/);
    assert.deepEqual(sequence, []); log.push('throws;sequence:empty');
  });
  await run('U04-effects-yield-with-error', async (root, log) => {
    const sequence = [];
    assert.throws(() => root.effect(function* () { yield () => sequence.push(1); throw new Error('test'); }), /test/);
    assert.deepEqual(sequence, [1]); log.push('throws;sequence:1');
  });
  await run('U05-effects-async-return-with-error', async (root, log) => {
    const sequence = [];
    const e = root.effect(async () => { throw new Error('test'); });
    assert.deepEqual(sequence, []); await assert.rejects(Promise.resolve(e));
    assert.deepEqual(sequence, []); log.push('rejects;sequence:empty');
  });
  return cases;
}

File: tests/Cordis.Conformance/Scenarios.cs
using Cordis;

namespace Cordis.Conformance;

public static partial class Scenarios
{
    public static IReadOnlyList<(string Id, Func<Task<string[]>> Run)> Cases { get; } =
    [
        ("C01-greeting-lifecycle", GreetingLifecycle),
        ("C02-effect-immediate-single-shot", ImmediateEffect),
        ("C03-reentrant-owner-disposal", ReentrantDisposal),
        ("C04-pending-owned-effect", PendingEffect),
        ("C05-dispose-before-load-checkpoint", BeforeCheckpoint),
        ("C06-async-setup-and-cleanup", AsyncSetup),
        ("C07-owner-joins-started-cleanup", OwnerJoins),
        ("C08-nested-effect-reverse-order", NestedEffects),
        ("C09-generator-setup-rollback", SetupRollback),
        ("C10-failed-group-does-not-starve-sibling", CleanupFailure),
        ("C11-reject-effect-during-unload", RejectDuringUnload),
        ("C12-apply-failure-cleanup", ApplyFailure),
        ("C13-loading-provider-invisible", ProviderVisibility),
        ("C14-inject-child-not-parent-restart", InjectChild),
        ("C15-pending-and-active-config-update", ConfigUpdate),
        ("C16-multiple-dependencies", MultipleDependencies),
        ("C17-owner-joins-disposing-child", DisposingChild),
        ("C18-stable-context-on-restart", StableContext),
        ("U01-effects-dispose-by-plugin", UpstreamDisposeByPlugin),
        ("U02-effects-dispose-manually", UpstreamDisposeManually),
        ("U03-effects-return-with-error", UpstreamReturnWithError),
        ("U04-effects-yield-with-error", UpstreamYieldWithError),
        ("U05-effects-async-return-with-error", UpstreamAsyncReturnWithError),
    ];

    private static Plugin<object?> Plugin(Action<Context> apply, params string[] inject) => new()
    {
        Inject = inject,
        Apply = (ctx, _) => apply(ctx),
    };

    private static async Task<string[]> InContext(Func<Context, List<string>, Task> run)
    {
        var trace = new List<string>();
        await using var root = new Context();
        await root.RunAsync(ctx => run(ctx, trace));
        return trace.ToArray();
    }

    private static Task<string[]> GreetingLifecycle() => InContext(async (ctx, log) =>
    {
        var a = new MessageSource();
        var b = new MessageSource();
        var plugin = new Plugin<string>
        {
            Inject = ["messages"],
            Apply = (c, prefix) =>
            {
                MessageSource source = c.Get<MessageSource>("messages")!;
                c.Effect(() => source.Subscribe(name => log.Add($"{prefix}, {name}!")));
            },
        };
        Fiber fiber = ctx.Plugin(plugin, "Hello");
        await fiber.WaitAsync();
        Equal(FiberState.Pending, fiber.State);
        log.Add("pending");
        EffectHandle first = ctx.Provide("messages", a);
        await fiber.WaitAsync();
        Equal(1, a.Count);
        a.Emit("Otto");
        fiber.Update("Hi");
        await fiber.WaitAsync();
        Equal(1, a.Count);
        a.Emit("Otto");
        await first.DisposeAsync();
        Equal(0, a.Count);
        Equal(FiberState.Pending, fiber.State);
        log.Add("provider-removed");
        EffectHandle second = ctx.Provide("messages", b);
        await fiber.WaitAsync();
        b.Emit("Otto");
        await fiber.DisposeAsync();
        Equal(0, b.Count);
        b.Emit("ignored");
        await second.DisposeAsync();
        log.Add("disposed");
        Equal("Hello, Otto!|Hi, Otto!", string.Join('|', log.Where(s => s.EndsWith('!')).Take(2)));
    });

    private static Task<string[]> ImmediateEffect() => InContext(async (ctx, log) =>
    {
        EffectHandle effect = ctx.Effect(() =>
        {
            log.Add("setup");
            return () => log.Add("cleanup");
        });
        log.Add("returned");
        await effect.Ready;
        await effect.DisposeAsync();
        await effect.DisposeAsync();
        log.Add("disposed-twice");
        Equal("setup|returned|cleanup|disposed-twice", string.Join('|', log));
    });

    private static Task<string[]> ReentrantDisposal() => InContext(async (ctx, log) =>
    {
        Task? stopping = null;
        ctx.Effect(() =>
        {
            log.Add("setup-start");
            stopping = ctx.Fiber.DisposeAsync().AsTask();
            log.Add("setup-end");
            return () => log.Add("cleanup");
        });
        log.Add("returned");
        await stopping!;
        Equal(FiberState.Active, ctx.Fiber.State); // Root disposal is restart, not deletion.
        log.Add("root-active");
    });

    private static Task<string[]> PendingEffect() => InContext(async (ctx, log) =>
    {
        Fiber fiber = ctx.Plugin(Plugin(_ => throw new Exception("must not apply"), "absent"));
        fiber.Context.Effect(() => { log.Add("setup"); return () => log.Add("cleanup"); });
        await fiber.WaitAsync();
        Equal(FiberState.Pending, fiber.State);
        await fiber.DisposeAsync();
        Equal(FiberState.Disposed, fiber.State);
        log.Add("disposed");
    });

    private static Task<string[]> BeforeCheckpoint() => InContext(async (ctx, log) =>
    {
        Fiber fiber = ctx.Plugin(Plugin(_ => log.Add("must-not-apply")));
        await fiber.DisposeAsync();
        Equal(FiberState.Disposed, fiber.State);
        Equal(0, log.Count);
        log.Add("disposed-before-apply");
    });

    private static Task<string[]> AsyncSetup() => InContext(async (ctx, log) =>
    {
        var setup = Gate();
        var cleanup = Gate();
        var cleanupEntered = Gate();
        EffectHandle effect = ctx.Effect(async () =>
        {
            log.Add("setup-start");
            await setup.Task;
            log.Add("setup-end");
            return (IAsyncDisposable)new AsyncCallback(async () =>
            {
                log.Add("cleanup-start");
                cleanupEntered.SetResult();
                await cleanup.Task;
                log.Add("cleanup-end");
            });
        });
        try
        {
            Task stopping = ctx.Fiber.DisposeAsync().AsTask();
            Equal(false, stopping.IsCompleted);
            log.Add("waits-for-setup");
            setup.SetResult();
            await cleanupEntered.Task;
            Equal(false, stopping.IsCompleted);
            log.Add("waits-for-cleanup");
            cleanup.SetResult();
            await stopping;
            await effect.Ready;
            log.Add("stopped");
        }
        finally
        {
            setup.TrySetResult();
            cleanup.TrySetResult();
        }
    });

    private static Task<string[]> OwnerJoins() => InContext(async (ctx, log) =>
    {
        var entered = Gate();
        var release = Gate();
        EffectHandle effect = ctx.Effect(() => new AsyncCallback(async () =>
        {
            log.Add("cleanup-start");
            entered.SetResult();
            await release.Task;
            log.Add("cleanup-end");
        }));
        try
        {
            Task first = effect.DisposeAsync().AsTask();
            await entered.Task;
            await effect.DisposeAsync();
            log.Add("second-public-call-returned");
            Task stop = ctx.Fiber.DisposeAsync().AsTask();
            Equal(false, stop.IsCompleted);
            log.Add("owner-waits");
            release.SetResult();
            await first;
            await stop;
            log.Add("stopped");
        }
        finally
        {
            release.TrySetResult();
        }
    });

    private static Task<string[]> NestedEffects() => InContext(async (ctx, log) =>
    {
        EffectHandle outer = ctx.Effect(() => new IAsyncDisposable[]
        {
            ctx.Effect(() => { log.Add("setup-a"); return () => log.Add("cleanup-a"); }),
            ctx.Effect(() => { log.Add("setup-b"); return () => log.Add("cleanup-b"); }),
        });
        await outer.DisposeAsync();
        await ctx.Fiber.DisposeAsync();
        Equal("setup-a|setup-b|cleanup-b|cleanup-a", string.Join('|', log));
    });

    private static Task<string[]> SetupRollback() => InContext(async (ctx, log) =>
    {
        IEnumerable<IAsyncDisposable> Setup()
        {
            yield return ctx.Effect(() => { log.Add("setup-a"); return () => log.Add("cleanup-a"); });
            yield return ctx.Effect(() => { log.Add("setup-b"); return () => log.Add("cleanup-b"); });
            throw new InvalidOperationException("setup-failed");
        }
        Throws<InvalidOperationException>(() => ctx.Effect(Setup));
        await ctx.Fiber.DisposeAsync();
        log.Add("failure-observed");
        Equal("setup-a|setup-b|cleanup-b|cleanup-a|failure-observed", string.Join('|', log));
    });

    private static async Task<string[]> CleanupFailure()
    {
        var errors = new List<Exception>();
        var log = new List<string>();
        await using var ctx = new Context(errors.Add);
        await ctx.RunAsync(async root =>
        {
            root.Effect(() => () => log.Add("sibling"));
            root.Effect(() => new IAsyncDisposable[]
            {
                new AsyncCallback(() => { log.Add("must-not-run"); return Task.CompletedTask; }),
                new AsyncCallback(() => { log.Add("failed-cleanup"); throw new InvalidOperationException("cleanup"); }),
            });
            await root.Fiber.DisposeAsync();
            Equal("failed-cleanup|sibling", string.Join('|', log));
            Equal(1, errors.Count);
            log.Add("one-diagnostic");
        });
        return log.ToArray();
    }

    private static Task<string[]> RejectDuringUnload() => InContext(async (ctx, log) =>
    {
        ctx.Effect(() => () =>
        {
            var error = Throws<CordisException>(() =>
                ctx.Effect(() => { log.Add("must-not-setup"); return () => { }; }));
            Equal("INACTIVE_EFFECT", error.Code);
            log.Add("rejected");
        });
        await ctx.Fiber.DisposeAsync();
        Equal(1, log.Count);
    });

    private static Task<string[]> ApplyFailure() => InContext(async (ctx, log) =>
    {
        Fiber fiber = ctx.Plugin(Plugin(c =>
        {
            c.Effect(() => { log.Add("setup"); return () => log.Add("cleanup"); });
            throw new InvalidOperationException("apply-failed");
        }));
        await ThrowsAsync<InvalidOperationException>(fiber.WaitAsync);
        Equal(FiberState.Failed, fiber.State);
        log.Add("failed");
        await fiber.DisposeAsync();
        Equal(FiberState.Disposed, fiber.State);
        log.Add("disposed");
    });

    private static Task<string[]> ProviderVisibility() => InContext(async (ctx, log) =>
    {
        var published = Gate();
        var release = Gate();
        var value = new object();
        Fiber consumer = ctx.Plugin(Plugin(_ => log.Add("consumer-active"), "service"));
        Fiber provider = ctx.Plugin(new Plugin<object?>
        {
            ApplyAsync = async (c, _) =>
            {
                c.Provide("service", value);
                published.SetResult();
                await release.Task;
            },
        });
        try
        {
            await published.Task;
            Equal(FiberState.Loading, provider.State);
            Equal<object?>(null, ctx.Get<object>("service"));
            Equal(true, ReferenceEquals(value, ctx.Get<object>("service", strict: false)));
            await consumer.WaitAsync();
            Equal(FiberState.Pending, consumer.State);
            log.Add("loading-not-visible");
            release.SetResult();
            await provider.WaitAsync();
            await consumer.WaitAsync();
            Equal(FiberState.Active, consumer.State);
            log.Add("ready");
        }
        finally
        {
            release.TrySetResult();
        }
    });

    private static Task<string[]> InjectChild() => InContext(async (ctx, log) =>
    {
        Fiber? child = null;
        int parents = 0;
        Fiber parent = ctx.Plugin(Plugin(c =>
        {
            parents++;
            child = c.Inject(["service"], cc =>
            {
                cc.Effect(() => { log.Add("child-setup"); return () => log.Add("child-cleanup"); });
            });
        }));
        await parent.WaitAsync();
        Equal(FiberState.Pending, child!.State);
        EffectHandle service = ctx.Provide("service", new object());
        await child.WaitAsync();
        await service.DisposeAsync();
        Equal(1, parents);
        Equal(FiberState.Active, parent.State);
        Equal(FiberState.Pending, child.State);
        log.Add("parent-unchanged");
        await parent.DisposeAsync();
        Equal(FiberState.Disposed, child.State);
        log.Add("child-disposed-with-parent");
    });

    private static Task<string[]> ConfigUpdate() => InContext(async (ctx, log) =>
    {
        int validations = 0;
        var plugin = new Plugin<string>
        {
            Inject = ["service"],
            Config = value =>
            {
                validations++;
                return value is string text && text == "good"
                    ? ConfigResult<string>.Success(text)
                    : ConfigResult<string>.Failure("expected good");
            },
            Apply = (_, config) => log.Add("apply:" + config),
        };
        Fiber fiber = ctx.Plugin(plugin, "bad");
        await fiber.WaitAsync();
        Equal(0, validations);
        fiber.Update("good");
        Equal(0, validations);
        log.Add("deferred-validation");
        ctx.Provide("service", new object());
        await fiber.WaitAsync();
        Throws<ConfigurationValidationException>(() => fiber.Update("bad"));
        Equal("good", (string)fiber.Config!);
        Equal("bad", (string)fiber.RawConfig!);
        Equal(FiberState.Active, fiber.State);
        log.Add("invalid-update-preserves-active-config");
        await ThrowsAsync<ConfigurationValidationException>(fiber.RestartAsync);
        Equal(FiberState.Failed, fiber.State);
        log.Add("restart-uses-raw-config");
        fiber.Update("good");
        await fiber.WaitAsync();
        Equal(FiberState.Active, fiber.State);
        log.Add("recovered");
    });

    private static Task<string[]> MultipleDependencies() => InContext(async (ctx, log) =>
    {
        int starts = 0;
        Fiber fiber = ctx.Plugin(Plugin(_ => starts++, "a", "b"));
        EffectHandle a = ctx.Provide("a", new object());
        await fiber.WaitAsync();
        Equal(FiberState.Pending, fiber.State);
        ctx.Provide("b", new object());
        await fiber.WaitAsync();
        Equal(1, starts);
        await a.DisposeAsync();
        Equal(FiberState.Pending, fiber.State);
        ctx.Provide("a", new object());
        await fiber.WaitAsync();
        Equal(2, starts);
        log.Add("both-required;two-activations");
    });

    private static Task<string[]> DisposingChild() => InContext(async (ctx, log) =>
    {
        var entered = Gate();
        var release = Gate();
        Fiber child = ctx.Plugin(Plugin(c => c.Effect(() => new AsyncCallback(async () =>
        {
            log.Add("child-cleanup-start");
            entered.SetResult();
            await release.Task;
            log.Add("child-cleanup-end");
        }))));
        try
        {
            await child.WaitAsync();
            Task first = child.DisposeAsync().AsTask();
            await entered.Task;
            Task rootStop = ctx.Fiber.DisposeAsync().AsTask();
            Equal(false, rootStop.IsCompleted);
            log.Add("root-waits");
            release.SetResult();
            await first;
            await rootStop;
            log.Add("stopped");
        }
        finally
        {
            release.TrySetResult();
        }
    });

    private static Task<string[]> StableContext() => InContext(async (ctx, log) =>
    {
        var seen = new List<Context>();
        Fiber fiber = ctx.Plugin(Plugin(seen.Add));
        await fiber.WaitAsync();
        long? uid = fiber.Uid;
        await fiber.RestartAsync();
        Equal(2, seen.Count);
        Equal(true, ReferenceEquals(seen[0], seen[1]));
        Equal(uid, fiber.Uid);
        log.Add("same-context;same-fiber;two-applies");
    });

    internal static void Equal<T>(T expected, T actual)
    {
        if (!EqualityComparer<T>.Default.Equals(expected, actual))
            throw new InvalidOperationException($"Expected '{expected}', actual '{actual}'.");
    }

    internal static T Throws<T>(Action operation) where T : Exception
    {
        try { operation(); }
        catch (T error) { return error; }
        throw new InvalidOperationException($"Expected {typeof(T).Name}.");
    }

    internal static async Task ThrowsAsync<T>(Func<Task> operation) where T : Exception
    {
        try { await operation(); }
        catch (T) { return; }
        throw new InvalidOperationException($"Expected {typeof(T).Name}.");
    }

    private static TaskCompletionSource Gate() => new(TaskCreationOptions.RunContinuationsAsynchronously);

    private sealed class AsyncCallback(Func<Task> callback) : IAsyncDisposable
    {
        public ValueTask DisposeAsync() => new(callback());
    }

    private sealed class Callback(Action action) : IDisposable
    {
        private Action? _action = action;
        public void Dispose() => Interlocked.Exchange(ref _action, null)?.Invoke();
    }

    private sealed class MessageSource
    {
        private readonly List<Action<string>> _handlers = [];
        public int Count => _handlers.Count;
        public IDisposable Subscribe(Action<string> handler)
        {
            _handlers.Add(handler);
            return new Callback(() => _handlers.Remove(handler));
        }
        public void Emit(string name)
        {
            foreach (Action<string> callback in _handlers.ToArray()) callback(name);
        }
    }
}
